src/process_util.rs

```diff
@@ -64,31 +64,57 @@ pub fn capture_fd3_status(mut cmd: Command)
     Ok((status, buf))
 }
 
-pub fn run_and_wait(cmd: &mut Command)
+pub fn run_and_wait(cmd: &mut Command, tty_fd: Option<i32>)
     -> Result<i32, String>
 {
     let mut trap = Trap::trap(&[SIGINT, SIGTERM, SIGCHLD]);
     info!("Running {:?}", cmd);
-    let child = try!(cmd.spawn().map_err(|e| format!("{}", e)));
+    let child = try!(cmd.spawn()
+                     .map_err(|e| format!("Error running {:?}: {}", cmd, e)));
+    let pgrp = unsafe { getpgrp() };
+    let tty_guard = match tty_fd {
+        Some(tty_fd) => Some(try!(TtyGuard::new(tty_fd, pgrp))),
+        None => None,
+    };
 
     for signal in trap.by_ref() {
         match signal {
             SIGINT => {
-                // SIGINT is usually a Ctrl+C so it's sent to whole process
-                // group, so we don't need to do anything special
+                // SIGINT is usually a Ctrl+C, if we trap it here
+                // child process hasn't controlling terminal,
+                // so we send the signal to the child process
                 debug!("Received SIGINT signal. Waiting process to stop..");
+                try!(child.signal(SIGINT)
+                     .map_err(|e| format!(
+                         "Error sending SIGINT to {:?}: {}", cmd, e)));
             }
             SIGTERM => {
                 // SIGTERM is usually sent to a specific process so we
                 // forward it to children
                 debug!("Received SIGTERM signal, propagating");
-                child.signal(SIGTERM).ok();
+                try!(child.signal(SIGTERM)
```

Comment:
Why do you think it's a fatal error? The error was ignored because we can't ensure that process is still alive.